pivot_wider() error: Assigned data `as.character(rep(values_from, each = 2))` must be compatible with existing data.

---
I am unable to execute pivot_wider() on a sample of data that used to work with sparklyr 1.7.5. Currently, with sparklyr 1.7.7, I am facing an error where pivot_wider() complains about size of the existing vs assigned data. Below is the error message:
Error:
! Assigned data `as.character(rep(values_from, each = 2))` must be compatible with existing data.
✖ Existing data has 2 rows.
✖ Assigned data has 4 rows.
ℹ Only vectors of size 1 are recycled.

As a reprex, I have created a dummy dataframe which I am executing on local R, Spark and SQLite. R and SQLite execute without errors and produce the desired output as shown below. However, sparklyr behaves differently given the number of input records.

``` r
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library(tidyr)
library(sparklyr)
#> 
#> Attaching package: 'sparklyr'
#> The following object is masked from 'package:stats':
#> 
#>     filter
library(DBI)
library(RSQLite)

# Create a dummy data frame
# Columns c1 and c2 identify the row uniquely after pivoting
# Column c3 is used for column generation
# Columns c4 and c5 have data that is widened
d <- data.frame(
  c1 = c(11, 11, 11)
  , c2 = c(21, 21, 21)
  , c3 = c(31, 32, 33)
  , c4 = c(41, 42, 43)
  , c5 = c(51, 52, 53)
)

# Pivoting data on local R - works fine all scenarios
d %>% head(1) %>% pivot_wider(id_cols = c(c1, c2), names_from = c3, values_from = c(c4, c5))
#> # A tibble: 1 × 4
#>      c1    c2 c4_31 c5_31
#>   <dbl> <dbl> <dbl> <dbl>
#> 1    11    21    41    51
d %>% head(2) %>% pivot_wider(id_cols = c(c1, c2), names_from = c3, values_from = c(c4, c5))
#> # A tibble: 1 × 6
#>      c1    c2 c4_31 c4_32 c5_31 c5_32
#>   <dbl> <dbl> <dbl> <dbl> <dbl> <dbl>
#> 1    11    21    41    42    51    52
d %>% head(3) %>% pivot_wider(id_cols = c(c1, c2), names_from = c3, values_from = c(c4, c5))
#> # A tibble: 1 × 8
#>      c1    c2 c4_31 c4_32 c4_33 c5_31 c5_32 c5_33
#>   <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl>
#> 1    11    21    41    42    43    51    52    53

spark_disconnect_all()
#> [1] 0
cn <- spark_connect("local")

# Pivoting data on Spark - works fine only for 2nd scenario with 2 records. 1st and 3rd scenarios fail.
d %>% head(1) %>% copy_to(cn, ., overwrite = T) %>% pivot_wider(id_cols = c(c1, c2), names_from = c3, values_from = c(c4, c5))
#> Error:
#> ! Assigned data `as.character(rep(values_from, each = 2))` must be compatible with existing data.
#> ✖ Existing data has 2 rows.
#> ✖ Assigned data has 4 rows.
#> ℹ Only vectors of size 1 are recycled.
d %>% head(2) %>% copy_to(cn, ., overwrite = T) %>% pivot_wider(id_cols = c(c1, c2), names_from = c3, values_from = c(c4, c5))
#> # Source: spark<?> [?? x 6]
#>      c1    c2 c4_31 c4_32 c5_31 c5_32
#>   <dbl> <dbl> <dbl> <dbl> <dbl> <dbl>
#> 1    11    21    41    42    51    52
d %>% head(3) %>% copy_to(cn, ., overwrite = T) %>% pivot_wider(id_cols = c(c1, c2), names_from = c3, values_from = c(c4, c5))
#> Error:
#> ! Assigned data `as.character(rep(values_from, each = 2))` must be compatible with existing data.
#> ✖ Existing data has 6 rows.
#> ✖ Assigned data has 4 rows.
#> ℹ Only vectors of size 1 are recycled.

# Pivoting data on non-Spark backend - works fine all scenarios
sqlc <- dbConnect(SQLite(), ":memory:")
d %>% head(1) %>% copy_to(sqlc, ., overwrite = T) %>% pivot_wider(id_cols = c(c1, c2), names_from = c3, values_from = c(c4, c5))
#> # Source:   SQL [1 x 4]
#> # Database: sqlite 3.38.5 [:memory:]
#>      c1    c2 c4_31 c5_31
#>   <dbl> <dbl> <dbl> <dbl>
#> 1    11    21    41    51
d %>% head(2) %>% copy_to(sqlc, ., overwrite = T) %>% pivot_wider(id_cols = c(c1, c2), names_from = c3, values_from = c(c4, c5))
#> # Source:   SQL [1 x 6]
#> # Database: sqlite 3.38.5 [:memory:]
#>      c1    c2 c4_31 c4_32 c5_31 c5_32
#>   <dbl> <dbl> <dbl> <dbl> <dbl> <dbl>
#> 1    11    21    41    42    51    52
d %>% head(3) %>% copy_to(sqlc, ., overwrite = T) %>% pivot_wider(id_cols = c(c1, c2), names_from = c3, values_from = c(c4, c5))
#> # Source:   SQL [1 x 8]
#> # Database: sqlite 3.38.5 [:memory:]
#>      c1    c2 c4_31 c4_32 c4_33 c5_31 c5_32 c5_33
#>   <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl>
#> 1    11    21    41    42    43    51    52    53
```

<sup>Created on 2022-09-05 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

<details style="margin-bottom:10px;">
<summary>
Session info
</summary>

``` r
sessioninfo::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value
#>  version  R version 4.0.2 (2020-06-22)
#>  os       Red Hat Enterprise Linux
#>  system   x86_64, linux-gnu
#>  ui       X11
#>  language (EN)
#>  collate  en_US.UTF-8
#>  ctype    en_US.UTF-8
#>  tz       America/Chicago
#>  date     2022-09-05
#>  pandoc   2.17.1.1 @ /opt/revr/rstudio-server/bin/quarto/bin/ (via rmarkdown)
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package     * version date (UTC) lib source
#>  askpass       1.1     2019-01-13 [2] CRAN (R 4.0.2)
#>  assertthat    0.2.1   2019-03-21 [2] CRAN (R 4.0.2)
#>  base64enc     0.1-3   2015-07-28 [2] CRAN (R 4.0.2)
#>  bit           4.0.4   2020-08-04 [2] CRAN (R 4.0.2)
#>  bit64         4.0.5   2020-08-30 [2] CRAN (R 4.0.2)
#>  blob          1.2.3   2022-04-10 [2] CRAN (R 4.0.2)
#>  cachem        1.0.6   2021-08-19 [2] CRAN (R 4.0.2)
#>  cli           3.3.0   2022-04-25 [2] CRAN (R 4.0.2)
#>  config        0.3.1   2020-12-17 [2] CRAN (R 4.0.2)
#>  crayon        1.5.1   2022-03-26 [2] CRAN (R 4.0.2)
#>  DBI         * 1.1.3   2022-06-18 [2] CRAN (R 4.0.2)
#>  dbplyr        2.2.1   2022-06-27 [2] CRAN (R 4.0.2)
#>  digest        0.6.29  2021-12-01 [2] CRAN (R 4.0.2)
#>  dplyr       * 1.0.9   2022-04-28 [2] CRAN (R 4.0.2)
#>  ellipsis      0.3.2   2021-04-29 [2] CRAN (R 4.0.2)
#>  evaluate      0.15    2022-02-18 [2] CRAN (R 4.0.2)
#>  fansi         1.0.3   2022-03-24 [2] CRAN (R 4.0.2)
#>  fastmap       1.1.0   2021-01-25 [2] CRAN (R 4.0.2)
#>  forge         0.2.0   2019-02-26 [2] CRAN (R 4.0.2)
#>  fs            1.5.2   2021-12-08 [2] CRAN (R 4.0.2)
#>  generics      0.1.3   2022-07-05 [2] CRAN (R 4.0.2)
#>  glue          1.6.2   2022-02-24 [2] CRAN (R 4.0.2)
#>  highr         0.9     2021-04-16 [2] CRAN (R 4.0.2)
#>  htmltools     0.5.2   2021-08-25 [2] CRAN (R 4.0.2)
#>  htmlwidgets   1.5.4   2021-09-08 [2] CRAN (R 4.0.2)
#>  httr          1.4.3   2022-05-04 [2] CRAN (R 4.0.2)
#>  jsonlite      1.8.0   2022-02-22 [2] CRAN (R 4.0.2)
#>  knitr         1.39    2022-04-26 [2] CRAN (R 4.0.2)
#>  lifecycle     1.0.1   2021-09-24 [2] CRAN (R 4.0.2)
#>  magrittr      2.0.3   2022-03-30 [2] CRAN (R 4.0.2)
#>  memoise       2.0.1   2021-11-26 [2] CRAN (R 4.0.2)
#>  openssl       2.0.2   2022-05-24 [2] CRAN (R 4.0.2)
#>  pillar        1.7.0   2022-02-01 [2] CRAN (R 4.0.2)
#>  pkgconfig     2.0.3   2019-09-22 [2] CRAN (R 4.0.2)
#>  purrr         0.3.4   2020-04-17 [2] CRAN (R 4.0.2)
#>  R.cache       0.15.0  2021-04-30 [2] CRAN (R 4.0.2)
#>  R.methodsS3   1.8.2   2022-06-13 [2] CRAN (R 4.0.2)
#>  R.oo          1.25.0  2022-06-12 [2] CRAN (R 4.0.2)
#>  R.utils       2.12.0  2022-06-28 [2] CRAN (R 4.0.2)
#>  r2d3          0.2.6   2022-02-28 [2] CRAN (R 4.0.2)
#>  R6            2.5.1   2021-08-19 [2] CRAN (R 4.0.2)
#>  Rcpp          1.0.9   2022-07-08 [2] CRAN (R 4.0.2)
#>  reprex        2.0.1   2021-08-05 [2] CRAN (R 4.0.2)
#>  rlang         1.0.4   2022-07-12 [2] CRAN (R 4.0.2)
#>  rmarkdown     2.14    2022-04-25 [2] CRAN (R 4.0.2)
#>  rprojroot     2.0.3   2022-04-02 [2] CRAN (R 4.0.2)
#>  RSQLite     * 2.2.14  2022-05-07 [2] CRAN (R 4.0.2)
#>  rstudioapi    0.13    2020-11-12 [2] CRAN (R 4.0.2)
#>  sessioninfo   1.2.2   2021-12-06 [2] CRAN (R 4.0.2)
#>  sparklyr    * 1.7.7   2022-06-07 [2] CRAN (R 4.0.2)
#>  stringi       1.7.8   2022-07-11 [2] CRAN (R 4.0.2)
#>  stringr       1.4.0   2019-02-10 [2] CRAN (R 4.0.2)
#>  styler        1.7.0   2022-03-13 [2] CRAN (R 4.0.2)
#>  tibble        3.1.7   2022-05-03 [2] CRAN (R 4.0.2)
#>  tidyr       * 1.2.0   2022-02-01 [2] CRAN (R 4.0.2)
#>  tidyselect    1.1.2   2022-02-21 [2] CRAN (R 4.0.2)
#>  utf8          1.2.2   2021-07-24 [2] CRAN (R 4.0.2)
#>  uuid          1.1-0   2022-04-19 [2] CRAN (R 4.0.2)
#>  vctrs         0.4.1   2022-04-13 [2] CRAN (R 4.0.2)
#>  withr         2.5.0   2022-03-03 [2] CRAN (R 4.0.2)
#>  xfun          0.31    2022-05-10 [2] CRAN (R 4.0.2)
#>  yaml          2.3.5   2022-02-21 [2] CRAN (R 4.0.2)
#>  
#> ──────────────────────────────────────────────────────────────────────────────
```

</details>

I had the same issue as the original post by @krprasangdas. However, your last proposed solution (using `spark_apply`) didn't work for me, I get a stage failure error.

Here's the example:

```{r}
library(sparklyr)
library(dplyr)

sc <- spark_connect(master = "local")

d <- data.frame(
  c1 = c(11, 11, 11)
  , c2 = c(21, 21, 21)
  , c3 = c(31, 32, 33)
  , c4 = c(41, 42, 43)
  , c5 = c(51, 52, 53)
) %>% copy_to(sc, ., "tmp", T)

d %>% 
  spark_apply(function(x) tidyr::pivot_wider(x, id_cols = c(c1, c2), names_from = c3, values_from = c(c4, c5)))

#> Error: org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 4.0 failed 1 times, most recent failure: Lost task 0.0 in stage 4.0 (TID 3) (127.0.0.1 executor driver): java.io.IOException: Cannot run program "Rscript" (in directory "C:\Users\alberto.agudo\AppData\Local\Temp\spark-c73193f6-ea89-4cde-a3a8-8a2a3763518f\userFiles-41544c19-f2a4-4ed7-a0a0-e0693389e94f"): CreateProcess error=2, The system cannot find the file specified
#>         at java.lang.ProcessBuilder.start(Unknown Source)
#>         at sparklyr.Rscript.init(rscript.scala:75)
#>         at sparklyr.WorkerApply$$anon$2.run(workerapply.scala:138)
#> Caused by: java.io.IOException: CreateProcess error=2, The system cannot find the file specified
#>         at java.lang.ProcessImpl.create(Native Method)
#>         at java.lang.ProcessImpl.<init>(Unknown Source)
#>         at java.lang.ProcessImpl.start(Unknown Source)
#>         ... 3 more
#> 
#> Driver stacktrace:
#>         at org.apache.spark.scheduler.DAGScheduler.failJobAndIndependentStages(DAGScheduler.scala:2672)
#>         at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2(DAGScheduler.scala:2608)
#>         at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2$adapted(DAGScheduler.scala:2607)
#>         at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
#>         at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
#>         at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
#>         at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:2607)
#>         at org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1(DAGScheduler.scala:1182)
#>         at org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1$adapted(DAGScheduler.scala:1182)
#>         at scala.Option.foreach(Option.scala:407)
#>         at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:1182)
#>         at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:2860)
#>         at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2802)
#>         at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2791)
#>         at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)
#>         at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:952)
#>         at org.apache.spark.SparkContext.runJob(SparkContext.scala:2228)
#>         at org.apache.spark.SparkContext.runJob(SparkContext.scala:2249)
#>         at org.apache.spark.SparkContext.runJob(SparkContext.scala:2268)
#>         at org.apache.spark.SparkContext.runJob(SparkContext.scala:2293)
#>         at org.apache.spark.rdd.RDD.$anonfun$collect$1(RDD.scala:1021)
#>         at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
#>         at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
#>         at org.apache.spark.rdd.RDD.withScope(RDD.scala:406)
#>         at org.apache.spark.rdd.RDD.collect(RDD.scala:1020)
#>         at org.apache.spark.sql.execution.SparkPlan.executeCollect(SparkPlan.scala:424)
#>         at org.apache.spark.sql.execution.adaptive.AdaptiveSparkPlanExec.$anonfun$executeCollect$1(AdaptiveSparkPlanExec.scala:340)
#>         at org.apache.spark.sql.execution.adaptive.AdaptiveSparkPlanExec.withFinalPlanUpdate(AdaptiveSparkPlanExec.scala:368)
#>         at org.apache.spark.sql.execution.adaptive.AdaptiveSparkPlanExec.executeCollect(AdaptiveSparkPlanExec.scala:340)
#>         at org.apache.spark.sql.Dataset.collectFromPlan(Dataset.scala:3868)
#>         at org.apache.spark.sql.Dataset.$anonfun$collect$1(Dataset.scala:3120)
#>         at org.apache.spark.sql.Dataset.$anonfun$withAction$2(Dataset.scala:3858)
#>         at org.apache.spark.sql.execution.QueryExecution$.withInternalError(QueryExecution.scala:510)
#>         at org.apache.spark.sql.Dataset.$anonfun$withAction$1(Dataset.scala:3856)
#>         at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$6(SQLExecution.scala:109)
#>         at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:169)
#>         at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$1(SQLExecution.scala:95)
#>         at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:779)
#>         at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:64)
#>         at org.apache.spark.sql.Dataset.withAction(Dataset.scala:3856)
#>         at org.apache.spark.sql.Dataset.collect(Dataset.scala:3120)
#>         at sparklyr.Utils$.collect(utils.scala:26)
#>         at sparklyr.Utils.collect(utils.scala)
#>         at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
#>         at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
#>         at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
#>         at java.lang.reflect.Method.invoke(Unknown Source)
#>         at sparklyr.Invoke.invoke(invoke.scala:161)
#>         at sparklyr.StreamHandler.handleMethodCall(stream.scala:141)
#>         at sparklyr.StreamHandler.read(stream.scala:62)
#>         at sparklyr.BackendHandler.$anonfun$channelRead0$1(handler.scala:60)
#>         at scala.util.control.Breaks.breakable(Breaks.scala:42)
#>         at sparklyr.BackendHandler.channelRead0(handler.scala:41)
#>         at sparklyr.BackendHandler.channelRead0(handler.scala:14)
#>         at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99)
#>         at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
#>         at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
#>         at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
#>         at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
#>         at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
#>         at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
#>         at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
#>         at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:327)
#>         at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:299)
#>         at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
#>         at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
#>         at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
#>         at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
#>         at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
#>         at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
#>         at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
#>         at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
#>         at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:722)
#>         at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:658)
#>         at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:584)
#>         at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:496)
#>         at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
#>         at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
#>         at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
#>         at java.lang.Thread.run(Unknown Source)
#> Caused by: java.io.IOException: Cannot run program "Rscript" (in directory "C:\Users\alberto.agudo\AppData\Local\Temp\spark-c73193f6-ea89-4cde-a3a8-8a2a3763518f\userFiles-41544c19-f2a4-4ed7-a0a0-e0693389e94f"): CreateProcess error=2, The system cannot find the file specified
#>         at java.lang.ProcessBuilder.start(Unknown Source)
#>         at sparklyr.Rscript.init(rscript.scala:75)
#>         at sparklyr.WorkerApply$$anon$2.run(workerap
```

---
## Session info

```{r}
> sessioninfo::session_info()
#> ─ Session info ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
#>  setting  value
#>  version  R version 4.2.2 (2022-10-31 ucrt)
#>  os       Windows 10 x64 (build 19045)
#>  system   x86_64, mingw32
#>  ui       RTerm
#>  language (EN)
#>  collate  English_United Kingdom.utf8
#>  ctype    English_United Kingdom.utf8
#>  tz       Europe/Paris
#>  date     2022-12-02
#>  pandoc   NA (via rmarkdown)
#> 
#> ─ Packages ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
#>  package     * version date (UTC) lib source
#>  askpass       1.1     2019-01-13 [1] CRAN (R 4.2.2)
#>  assertthat    0.2.1   2019-03-21 [1] CRAN (R 4.2.2)
#>  base64enc     0.1-3   2015-07-28 [1] CRAN (R 4.2.0)
#>  callr         3.7.3   2022-11-02 [1] CRAN (R 4.2.2)
#>  cli           3.4.1   2022-09-23 [1] CRAN (R 4.2.2)
#>  clipr         0.8.0   2022-02-22 [1] CRAN (R 4.2.2)
#>  config        0.3.1   2020-12-17 [1] CRAN (R 4.2.2)
#>  DBI           1.1.3   2022-06-18 [1] CRAN (R 4.2.2)
#>  dbplyr        2.2.1   2022-06-27 [1] CRAN (R 4.2.2)
#>  digest        0.6.30  2022-10-18 [1] CRAN (R 4.2.2)
#>  dplyr       * 1.0.10  2022-09-01 [1] CRAN (R 4.2.2)
#>  ellipsis      0.3.2   2021-04-29 [1] CRAN (R 4.2.2)
#>  evaluate      0.18    2022-11-07 [1] CRAN (R 4.2.2)
#>  fansi         1.0.3   2022-03-24 [1] CRAN (R 4.2.2)
#>  fastmap       1.1.0   2021-01-25 [1] CRAN (R 4.2.2)
#>  forge         0.2.0   2019-02-26 [1] CRAN (R 4.2.2)
#>  fs            1.5.2   2021-12-08 [1] CRAN (R 4.2.2)
#>  generics      0.1.3   2022-07-05 [1] CRAN (R 4.2.2)
#>  glue          1.6.2   2022-02-24 [1] CRAN (R 4.2.2)
#>  highr         0.9     2021-04-16 [1] CRAN (R 4.2.2)
#>  htmltools     0.5.3   2022-07-18 [1] CRAN (R 4.2.2)
#>  htmlwidgets   1.5.4   2021-09-08 [1] CRAN (R 4.2.2)
#>  httr          1.4.4   2022-08-17 [1] CRAN (R 4.2.2)
#>  jsonlite      1.8.3   2022-10-21 [1] CRAN (R 4.2.2)
#>  knitr         1.40    2022-08-24 [1] CRAN (R 4.2.2)
#>  lifecycle     1.0.3   2022-10-07 [1] CRAN (R 4.2.2)
#>  magrittr      2.0.3   2022-03-30 [1] CRAN (R 4.2.2)
#>  openssl       2.0.4   2022-10-17 [1] CRAN (R 4.2.2)
#>  pillar        1.8.1   2022-08-19 [1] CRAN (R 4.2.2)
#>  pkgconfig     2.0.3   2019-09-22 [1] CRAN (R 4.2.2)
#>  processx      3.8.0   2022-10-26 [1] CRAN (R 4.2.2)
#>  ps            1.7.2   2022-10-26 [1] CRAN (R 4.2.2)
#>  purrr         0.3.5   2022-10-06 [1] CRAN (R 4.2.2)
#>  R.cache       0.16.0  2022-07-21 [1] CRAN (R 4.2.2)
#>  R.methodsS3   1.8.2   2022-06-13 [1] CRAN (R 4.2.0)
#>  R.oo          1.25.0  2022-06-12 [1] CRAN (R 4.2.0)
#>  R.utils       2.12.2  2022-11-11 [1] CRAN (R 4.2.2)
#>  r2d3          0.2.6   2022-02-28 [1] CRAN (R 4.2.2)
#>  R6            2.5.1   2021-08-19 [1] CRAN (R 4.2.2)
#>  rappdirs      0.3.3   2021-01-31 [1] CRAN (R 4.2.2)
#>  reprex        2.0.2   2022-08-17 [1] CRAN (R 4.2.2)
#>  rlang         1.0.6   2022-09-24 [1] CRAN (R 4.2.2)
#>  rmarkdown     2.18    2022-11-09 [1] CRAN (R 4.2.2)
#>  rprojroot     2.0.3   2022-04-02 [1] CRAN (R 4.2.2)
#>  rstudioapi    0.14    2022-08-22 [1] CRAN (R 4.2.2)
#>  sessioninfo   1.2.2   2021-12-06 [1] CRAN (R 4.2.2)
#>  sparklyr    * 1.7.8   2022-08-16 [1] CRAN (R 4.2.2)
#>  styler        1.8.1   2022-11-07 [1] CRAN (R 4.2.2)
#>  tibble        3.1.8   2022-07-22 [1] CRAN (R 4.2.2)
#>  tidyr         1.2.1   2022-09-08 [1] CRAN (R 4.2.2)
#>  tidyselect    1.2.0   2022-10-10 [1] CRAN (R 4.2.2)
#>  utf8          1.2.2   2021-07-24 [1] CRAN (R 4.2.2)
#>  vctrs         0.5.1   2022-11-16 [1] CRAN (R 4.2.2)
#>  withr         2.5.0   2022-03-03 [1] CRAN (R 4.2.2)
#>  xfun          0.35    2022-11-16 [1] CRAN (R 4.2.2)
#>  yaml          2.3.6   2022-10-18 [1] CRAN (R 4.2.1)
#> 
#>  [1] C:/Program Files/R/R-4.2.2/library
```

@alberto-agudo Can you please retry by providing Rscript.exe path and NOT R.exe? The error suggests that Rscript is being looked for - Caused by: java.io.IOException: Cannot run program "Rscript"

--- a/DESCRIPTION
+++ b/DESCRIPTION
@@ -1,7 +1,7 @@
 Type: Package
 Package: sparklyr
 Title: R Interface to Apache Spark
-Version: 1.7.8
+Version: 1.7.8.9000
 Authors@R:
     c(person(given = "Javier",
              family = "Luraschi",
@@ -122,7 +122,7 @@ Suggests:
     parsnip,
     testthat
 Encoding: UTF-8
-RoxygenNote: 7.2.1
+RoxygenNote: 7.2.2
 SystemRequirements: Spark: 1.6.x, 2.x, or 3.x
 Collate:
     'spark_data_build_types.R'

--- a/NEWS.md
+++ b/NEWS.md
@@ -1,3 +1,9 @@
+# Sparklyr (dev)
+
+### Bug Fixes
+
+- Fixes `pivot_wider()` issues with simpler cases (#3289)
+
 # Sparklyr 1.7.8
 
 ### New features

--- a/R/tidyr_pivot_wider.R
+++ b/R/tidyr_pivot_wider.R
@@ -39,8 +39,8 @@ pivot_wider.tbl_spark <- function(data,
 
   spec <- sdf_build_wider_spec(
     data,
-    names_from = !! names_from,
-    values_from = !! values_from,
+    names_from = !!names_from,
+    values_from = !!values_from,
     names_prefix = names_prefix,
     names_sep = names_sep,
     names_glue = names_glue,
@@ -79,12 +79,12 @@ sdf_build_wider_spec <- function(data,
     collect()
 
   row_ids <- data %>%
-    select(!! enquo(names_from)) %>%
+    select(!!enquo(names_from)) %>%
     distinct() %>%
     collect() %>%
-    select(- dplyr::group_vars(data))
+    select(-dplyr::group_vars(data))
 
-  if(dim(row_ids)[2] == 1) {
+  if (dim(row_ids)[2] == 1) {
     row_names <- row_ids[1][[1]]
   } else {
     row_names <- transpose(row_ids) %>%
@@ -97,7 +97,7 @@ sdf_build_wider_spec <- function(data,
     out$.value <- values_from
   } else {
     out <- seq_along(values_from) %>%
-      map(~ out) %>%
+      map(~out) %>%
       reduce(rbind)
 
     out$.value <- as.character(rep(values_from, each = nrow(row_ids)))
@@ -114,8 +114,8 @@ sdf_build_wider_spec <- function(data,
   }
 
   if (names_sort) {
-    out <- arrange(out, .name)
-    }
+    out <- arrange(out, !!rlang::parse_expr(".name"))
+  }
 
   out
 }
@@ -301,8 +301,7 @@ sdf_pivot_wider <- function(data,
     dplyr::summarize %@% summarizers
 }
 
-.apply_pivot_wider_names_repair <- function(
-                                            data,
+.apply_pivot_wider_names_repair <- function(data,
                                             spec,
                                             spec_cols,
                                             key_vars,

--- a/R/tidyr_pivot_wider.R
+++ b/R/tidyr_pivot_wider.R
@@ -91,15 +91,21 @@ sdf_build_wider_spec <- function(data,
       map_chr(~ paste(.x, collapse = names_sep))
   }
 
-  out <- tibble::tibble(.name = paste0(names_prefix, row_names))
+  out <- tibble(.name = paste0(names_prefix, row_names))
 
   if (length(values_from) == 1) {
     out$.value <- values_from
   } else {
-    out <- reduce(map(seq_along(values_from), ~ out), rbind)
-    out$.value <- as.character(rep(values_from, each = 2))
+    out <- seq_along(values_from) %>%
+      map(~ out) %>%
+      reduce(rbind)
+
+    out$.value <- as.character(rep(values_from, each = nrow(row_ids)))
     out$.name <- paste0(out$.value, names_sep, out$.name)
-    row_ids <- reduce(map(seq_along(values_from), ~ row_ids), rbind)
+
+    row_ids <- seq_along(values_from) %>%
+      map(~row_ids) %>%
+      reduce(rbind)
   }
 
   out <- cbind(out, row_ids)
@@ -107,7 +113,9 @@ sdf_build_wider_spec <- function(data,
     out$.name <- as.character(glue::glue_data(out, names_glue))
   }
 
-  if (names_sort) out <- arrange(out, !! rlang::parse_expr(".name"))
+  if (names_sort) {
+    out <- arrange(out, .name)
+    }
 
   out
 }

--- a/tests/testthat/test-tidyr-pivot-wider.R
+++ b/tests/testthat/test-tidyr-pivot-wider.R
@@ -312,3 +312,39 @@ test_that("default `names_from` and `values_from` works as expected", {
 
   expect_equivalent(pv, tibble::tibble(x = 1, y = 2))
 })
+
+
+
+test_that("Simple cases work", {
+
+  d <- data.frame(
+    c1 = c(11, 11, 11),
+    c2 = c(21, 21, 21),
+    c3 = c(31, 32, 33),
+    c4 = c(41, 42, 43),
+    c5 = c(51, 52, 53)
+  )
+
+  ds <- copy_to(sc, d)
+
+  expect_equal(
+    collect(pivot_wider(head(ds, 1), id_cols = c(c1, c2), names_from = c3, values_from = c(c4, c5))),
+    tidyr::pivot_wider(head(d, 1), id_cols = c(c1, c2), names_from = c3, values_from = c(c4, c5))
+  )
+
+  expect_equal(
+    collect(pivot_wider(head(ds, 2), id_cols = c(c1, c2), names_from = c3, values_from = c(c4, c5))),
+    tidyr::pivot_wider(head(d, 2), id_cols = c(c1, c2), names_from = c3, values_from = c(c4, c5))
+  )
+
+  expect_equal(
+    collect(pivot_wider(ds, id_cols = c(c1, c2), names_from = c3, values_from = c(c4, c5))),
+    tidyr::pivot_wider(d, id_cols = c(c1, c2), names_from = c3, values_from = c(c4, c5))
+  )
+
+})
+
+
+
+
+

--- a/tests/testthat/test-tidyr-pivot-wider.R
+++ b/tests/testthat/test-tidyr-pivot-wider.R
@@ -316,6 +316,7 @@ test_that("default `names_from` and `values_from` works as expected", {
 
 
 test_that("Simple cases work", {
+  test_requires_version("2.3.0")
 
   d <- data.frame(
     c1 = c(11, 11, 11),

